hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/api/TokenServiceApiImpl.java

```diff
@@ -282,46 +280,60 @@ public void updateStorageMetadata(
+        chargePayer(payerAccount, amountToCharge);
+        rb.transactionFee(rb.transactionFee() + amountToCharge);
+        distributeToNetworkFundingAccounts(amountToCharge, rb);
     }
 
     @Override
-    public void chargeFees(@NonNull AccountID payer, @NonNull Fees fees, @NonNull final FeeRecordBuilder rb) {
+    public void chargeFees(
+            @NonNull AccountID payerId,
+            AccountID nodeAccountId,
+            @NonNull Fees fees,
+            @NonNull final FeeRecordBuilder rb) {
+        requireNonNull(rb);
+        requireNonNull(fees);
+        requireNonNull(payerId);
+        requireNonNull(nodeAccountId);
+
         // Note: these four accounts (payer, funding, staking reward, node reward) MUST exist for the transaction to be
         // valid and for fees to be processed. If any of them do not exist, the entire transaction will fail. There is
         // no conceivable way that these accounts *should* be null at this point.
-
-        // Record the total fee into the record builder
-        final var total = fees.totalFee();
-        rb.transactionFee(rb.transactionFee() + total);
-
-        // Charge the payer for the fees
-        chargePayer(payer, total);
-
-        // We may have a rounding error, so we will first remove the node and staking rewards from the total, and then
-        // whatever is left over goes to the funding account.
-        var balance = total;
-
-        // We only pay node and staking rewards if the feature is enabled
-        if (stakingConfig.isEnabled()) {
-            final var nodeReward = (long) ((stakingConfig.feesNodeRewardPercentage() / 100.0) * total);
-            balance -= nodeReward;
-            payNodeRewardAccount(nodeReward);
-
-            final var stakingReward = (long) ((stakingConfig.feesStakingRewardPercentage() / 100.0) * total);
-            balance -= stakingReward;
-            payStakingRewardAccount(stakingReward);
+        final var payerAccount = lookupAccount("Payer", payerId);
+        if (payerAccount.tinybarBalance() < fees.networkFee()) {
+            throw new IllegalArgumentException("Payer " + payerId
+                    + " (balance=" + payerAccount.tinybarBalance() + ") cannot afford network fee of "
+                    + fees.networkFee() + ", which should have been a due diligence failure");
+        }
+        if (fees.serviceFee() > 0 && payerAccount.tinybarBalance() < fees.totalFee()) {
+            throw new IllegalArgumentException("Payer " + payerId
+                    + " (balance=" + payerAccount.tinybarBalance() + ") cannot afford total fee of "
+                    + fees.totalFee() + ", which means service component should have been zeroed out");
+        }
+        // Prioritize network fee over node fee
+        final var chargeableNodeFee = Math.min(fees.nodeFee(), payerAccount.tinybarBalance() - fees.networkFee());
+        final var amountToCharge = fees.totalWithoutNodeFee() + chargeableNodeFee;
```

Comment:
In fact `fees.totalWithoutNodeFee() <= payerBalance` here because we have one of two cases:
  1. If `serviceFee > 0` then `payerBalance` must exceed `totalFee = amountToCharge`, or we would have thrown `IAE` above.
  2. If `serviceFee == 0` then `totalWithoutNodeFee == networkFee`; and from the preceding line `chargeableNodeFee <= payerBalance - networkFee`; so rearranging, `chargeableNodeFee + networkFee <= payerBalance`